pageserver/src/basebackup.rs

```diff
@@ -137,49 +137,52 @@ impl<'a> Basebackup<'a> {
     // Along with them also send PG_VERSION for each database.
     //
     fn add_relmap_file(&mut self, spcnode: u32, dbnode: u32) -> anyhow::Result<()> {
-        let img = self.timeline.get_page_at_lsn_nowait(
+        if let Ok(img) = self.timeline.get_page_at_lsn_nowait(
```

Comment:
Can you explain these changes? Why would we ever try to call `add_relmap_file` or `add_twophase_file` for something that doesn't exist in the repository?

Reply:
> We need this check because the cycle in `send_tarball()` iterates over all objects of this type and some of them can be stale (Unlinked) at this lsn.

`send_tarball()` uses `list_nonrels(lsn)` to get the list of objects. Why does `list_nonrels` return objects that don't in fact exist at that LSN?

Reply:
It's a bit tricky...
The short answer is because of the way we keep tombstones (RelationSizeEntry::Unlink) for non-rel relishes.

Let's say we have a db with dboid 100 and a FileNodeMap for this db.
When we save it, we create two objects in object store:
- the page itself  - key-value pair `(RelishTag::FileNodeMap(dboid:100),   pagedata)` at lsn 10
- the RelationSize tracker - key-value pair 
`(RelationMetadata(RelishTag::FileNodeMap(dboid:100)),   RelationSize(RelationSizeEntry::Size(1)))` at lsn 10

Then we drop the database and its FileNodeMap.
- insert a tombstone record - key-value pair 
`(RelationMetadata(RelishTag::FileNodeMap(dboid:100)),   RelationSize(RelationSizeEntry::Unlink))`  at lsn 20

`list_nonrels()` iterates through all non-rel page versions - ones with a key `RelishTag::FileNodeMap`, so it will return `(RelishTag::FileNodeMap(dboid:100))` at lsn 10.

Now we must check if it is stale or not. 
We can either call `get_relish_exists()` like it is done in `add_slru_segment()` or rely on internal check inside `get_page_at_lsn_nowait()`.

Reply:
> Maybe we should check if relish exists inside `list_nonrels()` after the `// visible in this snapshot` check.

Right :-).

> I think `list_nonrels()` needs some more refactoring. Almost all callers are only interested in one kind of object.
> So, probably it is worth splitting it into several functions `list_filenodemap`, `list_slru`, `list_twophase`.
> 
> What do you think?

Perhaps. I wouldn't worry about it, I don't think that's performance critical. But no objections if you want to do it now.

Reply:
Done. 
It required using `object_repository::{ObjectValue, RelationSizeEntry};` from `rocksdb_storage`.
I suppose it violates abstraction layers, but I don't feel like detangling it even further...